File: ios/Runner/Decart/DecartVideoPlatformView.swift
import Flutter
import LiveKit
import UIKit

enum DecartVideoViewType {
  /// Keep in sync with `DecartVideoView.viewType` on the Dart side.
  static let id = "livelook/decart_video"
}

/// Builds one `DecartVideoPlatformView` per `UiKitView` in the widget tree.
final class DecartVideoViewFactory: NSObject, FlutterPlatformViewFactory {
  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    let raw = (args as? [String: Any])?["source"] as? String ?? ""
    let source = DecartVideoSource(rawValue: raw) ?? .local
    return DecartVideoPlatformView(source: source)
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    FlutterStandardMessageCodec.sharedInstance()
  }
}

/// Renders one Decart video feed inside the Flutter widget tree.
///
/// The track is not owned here: `DecartSessionBridge` hands one over whenever
/// the session produces or drops it, so this view survives connect/disconnect
/// without being rebuilt.
final class DecartVideoPlatformView: NSObject, FlutterPlatformView {
  let source: DecartVideoSource

  private let videoView = VideoView()

  init(source: DecartVideoSource) {
    self.source = source
    super.init()

    // Flutter creates platform views on the platform thread, which is the main
    // thread — so the main-actor work below is safe to assume rather than hop
    // to, and the view is configured before its first frame.
    MainActor.assumeIsolated {
      videoView.backgroundColor = .black
      // Fill rather than letterbox: this is a mirror, and black bars around a
      // person read as a bug.
      videoView.layoutMode = .fill
      // The capture track already carries MirroringVideoProcessor; mirroring
      // again here would undo it.
      videoView.mirrorMode = .off

      DecartSessionBridge.shared.register(self)
    }
  }

  func view() -> UIView {
    videoView
  }

  /// Called by the bridge on the main actor.
  @MainActor
  func attach(_ track: VideoTrack?) {
    videoView.track = track
  }

  deinit {
    let view = videoView
    Task { @MainActor in
      view.track = nil
    }
  }
}

File: android/app/src/main/kotlin/com/livelook/app/decart/DecartVideoPlatformView.kt
package com.livelook.app.decart

import android.content.Context
import android.graphics.Color
import android.view.View
import android.widget.FrameLayout
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import io.livekit.android.renderer.TextureViewRenderer
import io.livekit.android.room.Room
import io.livekit.android.room.track.VideoTrack
import livekit.org.webrtc.RendererCommon

object DecartVideoViewType {
    /** Keep in sync with `DecartVideoView.viewType` on the Dart side. */
    const val ID = "livelook/decart_video"
}

/** Builds one [DecartVideoPlatformView] per `AndroidView` in the widget tree. */
class DecartVideoViewFactory : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        @Suppress("UNCHECKED_CAST")
        val params = args as? Map<String, Any?>
        val source = DecartVideoSource.fromWire(params?.get("source") as? String)
        return DecartVideoPlatformView(context, source)
    }
}

/**
 * Renders one Decart video feed inside the Flutter widget tree.
 *
 * The track is not owned here: [DecartSessionBridge] hands one over whenever
 * the session produces or drops it.
 *
 * A `TextureViewRenderer` is bound to the `EglBase` of the Room it was
 * initialised against, and re-initialising one against a different `EglBase`
 * is not reliable — so a new Room means a brand new renderer rather than a
 * rebind. Ending a session creates a fresh camera stream, and with it a fresh
 * Room, so this path runs every time.
 */
class DecartVideoPlatformView(
    private val context: Context,
    val source: DecartVideoSource,
) : PlatformView {

    private val container = FrameLayout(context).apply {
        setBackgroundColor(Color.BLACK)
    }

    private var renderer: TextureViewRenderer? = null
    private var boundTrack: VideoTrack? = null
    private var boundRoom: Room? = null

    init {
        DecartSessionBridge.register(this)
    }

    override fun getView(): View = container

    fun attach(track: VideoTrack?, room: Room?) {
        if (room == null) {
            releaseRenderer()
            return
        }

        if (room !== boundRoom) {
            releaseRenderer()
            createRenderer(room)
        }

        val active = renderer ?: return
        if (track === boundTrack) return

        boundTrack?.runCatching { removeRenderer(active) }
        track?.addRenderer(active)
        boundTrack = track
    }

    private fun createRenderer(room: Room) {
        val created = TextureViewRenderer(context)
        created.init(room.lkObjects.eglBase.eglBaseContext, null)
        created.setEnableHardwareScaler(true)
        // Fill rather than letterbox: this is a mirror, and black bars around
        // a person read as a bug.
        created.setScalingType(RendererCommon.ScalingType.SCALE_ASPECT_FILL)
        // The capture stream is already mirrored by MirrorMode.AUTO; mirroring
        // again here would undo it.
        created.setMirror(false)

        container.addView(
            created,
            FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT,
            ),
        )

        renderer = created
        boundRoom = room
        boundTrack = null
    }

    private fun releaseRenderer() {
        val active = renderer ?: return
        boundTrack?.runCatching { removeRenderer(active) }
        boundTrack = null
        container.removeView(active)
        runCatching { active.release() }
        renderer = null
        boundRoom = null
    }

    override fun dispose() {
        DecartSessionBridge.unregister(this)
        releaseRenderer()
    }
}
